File: cql-adapters/java-driver-3/src/main/java/com/scylladb/latte/SessionManager.java
package com.scylladb.latte;

import com.datastax.driver.core.Cluster;
import com.datastax.driver.core.PoolingOptions;
import com.datastax.driver.core.PreparedStatement;
import com.datastax.driver.core.Session;
import com.datastax.driver.core.SocketOptions;
import com.datastax.driver.core.policies.DCAwareRoundRobinPolicy;
import com.datastax.driver.core.policies.ConstantSpeculativeExecutionPolicy;
import java.net.InetSocketAddress;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Manages database sessions and prepared statement caches for Java driver 3.x.
 *
 * <p>This class is responsible for:
 *
 * <ul>
 *   <li>Creating and configuring CQL sessions based on connection parameters
 *   <li>Assigning unique session IDs for client reference
 *   <li>Caching prepared statements per session
 *   <li>Providing session lifecycle management (create, get, close)
 * </ul>
 *
 * <p>Sessions are stored in a thread-safe map and can be accessed by their unique session ID.
 *
 * <h2>Configuration Parameters</h2>
 *
 * <p>The following parameters can be passed to {@link #createSession(Map)}:
 *
 * <ul>
 *   <li>{@code contact_points} - Comma-separated list of host:port pairs
 *   <li>{@code keyspace} - Default keyspace for the session
 *   <li>{@code datacenter} - Local datacenter for load balancing
 *   <li>{@code username} / {@code password} - Authentication credentials
 *   <li>{@code request_timeout_ms} - Request timeout in milliseconds
 *   <li>{@code connect_timeout_ms} - Connection timeout in milliseconds
 *   <li>{@code connections_per_shard} - Pool size per host
 *   <li>{@code speculative_execution} - Enable speculative execution ("true"/"false")
 *   <li>{@code speculative_execution_max} - Max speculative executions (default: 2)
 *   <li>{@code speculative_execution_delay_ms} - Delay before speculative execution in ms (default:
 *       500)
 * </ul>
 *
 * <h2>Speculative Execution</h2>
 *
 * <p>When enabled, speculative execution sends the same request to multiple nodes simultaneously,
 * using the first response received. This can significantly reduce tail latencies at the cost of
 * additional load on the cluster.
 *
 * <p>Example configuration:
 *
 * <pre>
 * speculative_execution=true
 * speculative_execution_max=2
 * speculative_execution_delay_ms=100
 * </pre>
 */
public class SessionManager {
  private static final Logger logger = LoggerFactory.getLogger(SessionManager.class);

  private final AtomicLong sessionIdGenerator = new AtomicLong(1);
  private final ConcurrentHashMap<Long, SessionEntry> sessions = new ConcurrentHashMap<>();

  public SessionManager() {
  }

  /**
   * Create a new session with the given parameters.
   *
   * @param params connection parameters (see class documentation for supported keys)
   * @return the unique session ID for this session
   * @throws RuntimeException if the session cannot be created
   */
  public long createSession(Map<String, String> params) {
    long sessionId = sessionIdGenerator.getAndIncrement();

    // Get contact points from params
    String contactPoints = params.getOrDefault("contact_points", "");

    Cluster.Builder builder = Cluster.builder();

    // Parse contact points
    for (String cp : contactPoints.split(",")) {
      cp = cp.trim();
      if (cp.isEmpty()) continue;

      String host;
      int port = 9042;
      if (cp.contains(":")) {
        String[] parts = cp.split(":");
        host = parts[0];
        port = Integer.parseInt(parts[1]);
      } else {
        host = cp;
      }
      builder.addContactPointsWithPorts(new InetSocketAddress(host, port));
    }

    // Configure socket options (timeouts)
    SocketOptions socketOptions = new SocketOptions();
    socketOptions.setConnectTimeoutMillis(getConnectTimeoutMs(params));
    socketOptions.setReadTimeoutMillis(getRequestTimeoutMs(params));
    builder.withSocketOptions(socketOptions);

    // Configure pooling options
    int connectionsPerHost = getConnectionsPerShard(params);
    PoolingOptions poolingOptions = new PoolingOptions();
    poolingOptions.setCoreConnectionsPerHost(
        com.datastax.driver.core.HostDistance.LOCAL, connectionsPerHost);
    poolingOptions.setMaxConnectionsPerHost(
        com.datastax.driver.core.HostDistance.LOCAL, connectionsPerHost);
    poolingOptions.setCoreConnectionsPerHost(
        com.datastax.driver.core.HostDistance.REMOTE, connectionsPerHost);
    poolingOptions.setMaxConnectionsPerHost(
        com.datastax.driver.core.HostDistance.REMOTE, connectionsPerHost);
    builder.withPoolingOptions(poolingOptions);

    // Configure load balancing policy with datacenter
    String datacenter = params.get("datacenter");
    if (datacenter != null && !datacenter.isEmpty()) {
      builder.withLoadBalancingPolicy(
          DCAwareRoundRobinPolicy.builder().withLocalDc(datacenter).build());
    } else {
      // Use a default datacenter for local development
      builder.withLoadBalancingPolicy(
          DCAwareRoundRobinPolicy.builder().withLocalDc("datacenter1").build());
    }

    // Configure speculative execution if enabled
    if (isSpeculativeExecutionEnabled(params)) {
      int maxSpeculative = getSpeculativeExecutionMax(params);
      int speculativeDelayMs = getSpeculativeExecutionDelayMs(params);

      builder.withSpeculativeExecutionPolicy(
          new ConstantSpeculativeExecutionPolicy(speculativeDelayMs, maxSpeculative));

      logger.info(
          "Speculative execution enabled: max={}, delay={}ms", maxSpeculative, speculativeDelayMs);
    }

    // Set authentication if provided
    String username = params.get("username");
    String password = params.get("password");
    if (username != null && password != null) {
      builder.withCredentials(username, password);
    }

    Cluster cluster = builder.build();

    // Connect to keyspace if provided
    String keyspace = params.get("keyspace");
    Session session;
    if (keyspace != null && !keyspace.isEmpty()) {
      session = cluster.connect(keyspace);
    } else {
      session = cluster.connect();
    }

    sessions.put(sessionId, new SessionEntry(cluster, session));

    logger.info(
        "Created session {} with contact points: {}, keyspace: {}",
        sessionId,
        contactPoints,
        keyspace);

    return sessionId;
  }

  /**
   * Get a session by ID.
   *
   * @param sessionId the session ID returned from {@link #createSession(Map)}
   * @return the session entry containing the CQL session and prepared statement cache
   * @throws Protocol.ProtocolException if the session does not exist
   */
  public SessionEntry getSession(long sessionId) {
    SessionEntry entry = sessions.get(sessionId);
    if (entry == null) {
      // Simple error message to avoid allocation in hot path; detailed info available via logging
      throw new Protocol.ProtocolException("Session not found: " + sessionId);
    }
    return entry;
  }

  /** Close a session. */
  public void closeSession(long sessionId) {
    SessionEntry entry = sessions.remove(sessionId);
    if (entry != null) {
      entry.session().close();
      entry.cluster().close();
      logger.info("Closed session {}", sessionId);
    }
  }

  /** Close all sessions. */
  public void closeAll() {
    for (Map.Entry<Long, SessionEntry> entry : sessions.entrySet()) {
      entry.getValue().session().close();
      entry.getValue().cluster().close();
    }
    sessions.clear();
    logger.info("Closed all sessions");
  }

  private int getRequestTimeoutMs(Map<String, String> params) {
    String value = params.get("request_timeout_ms");
    if (value != null) {
      return Integer.parseInt(value);
    }
    return 12000; // 12 seconds default
  }

  private int getConnectTimeoutMs(Map<String, String> params) {
    String value = params.get("connect_timeout_ms");
    if (value != null) {
      return Integer.parseInt(value);
    }
    return 5000; // 5 seconds default
  }

  private int getConnectionsPerShard(Map<String, String> params) {
    String value = params.get("connections_per_shard");
    if (value != null) {
      return Integer.parseInt(value);
    }
    return 1;
  }

  private boolean isSpeculativeExecutionEnabled(Map<String, String> params) {
    String value = params.get("speculative_execution");
    return "true".equalsIgnoreCase(value);
  }

  private int getSpeculativeExecutionMax(Map<String, String> params) {
    String value = params.get("speculative_execution_max");
    if (value != null) {
      return Integer.parseInt(value);
    }
    return 2; // Default: up to 2 speculative executions
  }

  private int getSpeculativeExecutionDelayMs(Map<String, String> params) {
    String value = params.get("speculative_execution_delay_ms");
    if (value != null) {
      return Integer.parseInt(value);
    }
    return 500; // Default: 500ms before speculative execution
  }

  /**
   * A session entry with its cluster, session, and prepared statement cache.
   *
   * <p>Note: In driver 3.x, we need to keep both Cluster and Session to properly manage lifecycle.
   */
  public static final class SessionEntry {
    private final Cluster cluster;
    private final Session session;
    private final ConcurrentHashMap<String, PreparedStatement> preparedStatements;

    /**
     * Creates a session entry with an empty prepared statement cache.
     *
     * @param cluster the cluster
     * @param session the session
     */
    public SessionEntry(Cluster cluster, Session session) {
      this.cluster = cluster;
      this.session = session;
      this.preparedStatements = new ConcurrentHashMap<>();
    }

    public Cluster cluster() {
      return cluster;
    }

    public Session session() {
      return session;
    }

    public ConcurrentHashMap<String, PreparedStatement> preparedStatements() {
      return preparedStatements;
    }

    /**
     * Get a prepared statement from the cache.
     *
     * @param key the statement key
     * @return the prepared statement, or null if not found
     */
    public PreparedStatement getPrepared(String key) {
      return preparedStatements.get(key);
    }

    /**
     * Cache a prepared statement.
     *
     * @param key the statement key
     * @param ps the prepared statement to cache
     */
    public void putPrepared(String key, PreparedStatement ps) {
      preparedStatements.put(key, ps);
    }
  }
}

File: cql-adapters/csharp-driver/src/SessionManager.cs
using System.Collections.Concurrent;
using Cassandra;
using Microsoft.Extensions.Logging;

namespace LatteDriver;

/// <summary>
/// Caches prepared statements for a driver session.
/// </summary>
public class CachedPrepared
{
    public PreparedStatement Statement { get; }
    public ColumnDesc[]? BindColumns { get; }

    public CachedPrepared(PreparedStatement statement)
    {
        Statement = statement;
        BindColumns = statement.Variables?.Columns;
    }
}

/// <summary>
/// Wraps a Cassandra session with prepared statement caching.
/// </summary>
public class DriverSession : IDisposable
{
    private readonly ISession _session;
    private readonly ConcurrentDictionary<string, CachedPrepared> _preparedCache = new();
    private readonly ILogger _logger;
    private bool _disposed;

    public DriverSession(ISession session, ILogger logger)
    {
        _session = session;
        _logger = logger;
    }

    public ISession Session => _session;

    public CachedPrepared? GetPrepared(string statementKey)
    {
        _preparedCache.TryGetValue(statementKey, out var cached);
        return cached;
    }

    public async Task<CachedPrepared> PrepareAsync(string query, string statementKey)
    {
        var prepared = await _session.PrepareAsync(query);
        var cached = new CachedPrepared(prepared);
        _preparedCache[statementKey] = cached;
        _logger.LogDebug("Prepared statement '{Key}' for query: {Query}", statementKey, query);
        return cached;
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;
        _session.Dispose();
    }
}

/// <summary>
/// Manages driver sessions and their lifecycle.
/// </summary>
public class SessionRegistry : IDisposable
{
    private readonly ConcurrentDictionary<ulong, DriverSession> _sessions = new();
    private ulong _nextSessionId = 1;
    private readonly ILoggerFactory _loggerFactory;
    private readonly ILogger _logger;
    private bool _disposed;

    public SessionRegistry(ILoggerFactory loggerFactory)
    {
        _loggerFactory = loggerFactory;
        _logger = loggerFactory.CreateLogger<SessionRegistry>();
    }

    public DriverSession? Get(ulong sessionId)
    {
        _sessions.TryGetValue(sessionId, out var session);
        return session;
    }

    public async Task<ulong> CreateSessionAsync(Dictionary<string, string> parameters)
    {
        var sessionId = Interlocked.Increment(ref _nextSessionId);

        // Parse parameters
        var contactPoints = parameters.GetValueOrDefault("contact_points", "");
        var keyspace = parameters.GetValueOrDefault("keyspace");
        var username = parameters.GetValueOrDefault("username");
        var password = parameters.GetValueOrDefault("password");
        var datacenter = parameters.GetValueOrDefault("datacenter");
        var connectionsPerHost = 2;
        if (parameters.TryGetValue("connections_per_shard", out var connStr) && int.TryParse(connStr, out var connVal))
            connectionsPerHost = connVal;

        int requestTimeoutMs = 12000; // 12 seconds default
        if (parameters.TryGetValue("request_timeout_ms", out var timeoutStr) && int.TryParse(timeoutStr, out var timeoutVal))
            requestTimeoutMs = timeoutVal;

        _logger.LogInformation("Creating session {Id} to {ContactPoints} (keyspace: {Keyspace})",
            sessionId, contactPoints, keyspace ?? "none");

        // Build cluster
        var builder = Cluster.Builder();

        // Parse and add contact points
        foreach (var cp in contactPoints.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries))
        {
            var parts = cp.Split(':');
            var host = parts[0];
            var port = parts.Length > 1 && int.TryParse(parts[1], out var p) ? p : 9042;
            builder.AddContactPoint(host).WithPort(port);
        }

        // Authentication
        if (!string.IsNullOrEmpty(username))
        {
            builder.WithCredentials(username, password ?? "");
        }

        // Connection pool
        builder.WithPoolingOptions(
            new PoolingOptions()
                .SetCoreConnectionsPerHost(HostDistance.Local, connectionsPerHost)
                .SetMaxConnectionsPerHost(HostDistance.Local, connectionsPerHost * 2));

        // Socket options
        builder.WithSocketOptions(
            new SocketOptions()
                .SetReadTimeoutMillis(requestTimeoutMs)
                .SetConnectTimeoutMillis(requestTimeoutMs));

        // Query options
        var consistency = ConsistencyLevel.LocalQuorum;
        if (parameters.TryGetValue("consistency", out var consStr))
        {
            consistency = ParseConsistency(consStr);
        }
        builder.WithQueryOptions(new QueryOptions().SetConsistencyLevel(consistency));

        // Load balancing
        ILoadBalancingPolicy loadBalancing;
        if (!string.IsNullOrEmpty(datacenter))
        {
            loadBalancing = new TokenAwarePolicy(new DCAwareRoundRobinPolicy(datacenter));
        }
        else
        {
            loadBalancing = new TokenAwarePolicy(new RoundRobinPolicy());
        }
        builder.WithLoadBalancingPolicy(loadBalancing);

        // Build and connect
        var cluster = builder.Build();
        ISession session;
        if (!string.IsNullOrEmpty(keyspace))
        {
            session = await cluster.ConnectAsync(keyspace);
        }
        else
        {
            session = await cluster.ConnectAsync();
        }

        var driverSession = new DriverSession(session, _loggerFactory.CreateLogger<DriverSession>());
        _sessions[sessionId] = driverSession;

        _logger.LogInformation("Session {Id} created successfully", sessionId);
        return sessionId;
    }

    private static ConsistencyLevel ParseConsistency(string value)
    {
        if (value.Equals("ANY", StringComparison.OrdinalIgnoreCase))
            return ConsistencyLevel.Any;
        if (value.Equals("ONE", StringComparison.OrdinalIgnoreCase))
            return ConsistencyLevel.One;
        if (value.Equals("TWO", StringComparison.OrdinalIgnoreCase))
            return ConsistencyLevel.Two;
        if (value.Equals("THREE", StringComparison.OrdinalIgnoreCase))
            return ConsistencyLevel.Three;
        if (value.Equals("QUORUM", StringComparison.OrdinalIgnoreCase))
            return ConsistencyLevel.Quorum;
        if (value.Equals("ALL", StringComparison.OrdinalIgnoreCase))
            return ConsistencyLevel.All;
        if (value.Equals("LOCAL_QUORUM", StringComparison.OrdinalIgnoreCase))
            return ConsistencyLevel.LocalQuorum;
        if (value.Equals("EACH_QUORUM", StringComparison.OrdinalIgnoreCase))
            return ConsistencyLevel.EachQuorum;
        if (value.Equals("LOCAL_ONE", StringComparison.OrdinalIgnoreCase))
            return ConsistencyLevel.LocalOne;
        if (value.Equals("SERIAL", StringComparison.OrdinalIgnoreCase))
            return ConsistencyLevel.Serial;
        if (value.Equals("LOCAL_SERIAL", StringComparison.OrdinalIgnoreCase))
            return ConsistencyLevel.LocalSerial;
        return ConsistencyLevel.LocalQuorum;
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;

        foreach (var session in _sessions.Values)
        {
            session.Dispose();
        }
        _sessions.Clear();
    }
}
